resourcemanager/pool/spool/dynamic_worker.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spool
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/resourcemanager/pool"
+	"github.com/pingcap/tidb/resourcemanager/poolmanager"
+)
+
+// RunWithDynamicalConcurrency runs a function in the pool with dynamically concurrency.
+func RunWithDynamicalConcurrency[T any, W Worker[T]](ctx context.Context, p *Pool, create func() W, least, concurrency uint32) (*DynamicConcurrencyHandle[T, W], error) {
+	if p.isStop.Load() {
+		return nil, pool.ErrPoolClosed
+	}
+	if least > concurrency {
+		least = concurrency
+	}
+	conc, run := p.checkAndAddRunning(least)
+	if !run {
+		return nil, pool.ErrPoolOverload
+	}
+	taskID := p.GenTaskID()
+	exitCh := make(chan struct{})
+	// hack: use a closed channel to avoid blocking runTask when tuning.
+	close(exitCh)
+	meta := poolmanager.NewMeta(taskID, exitCh, nil, conc)
+	p.taskManager.RegisterTask(meta)
+	ctx, cancel := context.WithCancel(ctx)
+	handle := &DynamicConcurrencyHandle[T, W]{
+		ctx:    ctx,
+		cancel: cancel,
+		max:    int32(concurrency),
+		taskID: taskID,
+		p:      p,
+		taskCh: make(chan T, concurrency),
+		create: create,
+	}
+	handle.cap.Store(conc)
+	for n := int32(0); n < conc; n++ {
+		p.run(handle.run)
+	}
+	return handle, nil
+}
+
+// Worker interface.
+type Worker[task any] interface {
+	HandleTask(context.Context, task)
+	Close()
+}
+
+// DynamicConcurrencyHandle is the handle of dynamic concurrency worker.
+type DynamicConcurrencyHandle[T any, W Worker[T]] struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	running atomic.Int32
+	cap     atomic.Int32
+	max     int32
+	taskID  uint64
+	p       *Pool
+	taskCh  chan T
+	create  func() W
+	wg      sync.WaitGroup
+}
+
+// Send a function to the pool, spawn new worker if required.
+func (d *DynamicConcurrencyHandle[T, W]) Send(task T) {
+	trySpawn := len(d.taskCh) > 0 || d.running.Load() == 0
+	d.taskCh <- task
+	if trySpawn {
+		capacity := d.cap.Load()
+		var swapped bool
+		for {
+			if capacity > d.max {
+				break
+			}
+			swapped = d.cap.CompareAndSwap(capacity, capacity+1)
+			if swapped {
+				break
+			}
+			capacity = d.cap.Load()
+		}
+		if swapped && capacity <= d.max {
```

Comment:
`capacity <= d.max` seems to be always true. Is this an intentional safety check or did I miss anything?